backupd/tests/server_tests.rs

```diff
@@ -0,0 +1,169 @@
+use std::cmp;
+use std::collections::{HashMap, HashSet};
+use std::io::{Read, Write, Seek, SeekFrom, Result as IoResult};
+use std::sync::{Arc, Mutex};
+
+use backupd::storage::{StorageManager, FileLen};
+use backupd::server::BaacupImpl;
+
+use backuplib::rpc::{Baacup, FileMetadata, FileChunk};
+
+#[derive(Debug, Clone)]
+pub struct InMemoryStorage {
+    map_mutex: Arc<Mutex<HashMap<String, Arc<Mutex<Vec<u8>>>>>>,
+}
+
+impl InMemoryStorage {
+    pub fn new() -> InMemoryStorage {
+        InMemoryStorage {
+            map_mutex: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+impl<'a> StorageManager<'a> for InMemoryStorage {
+    type File = InMemoryFile;
+
+    fn create_storage(&'a self, path: String) -> Result<InMemoryFile, String> {
+        let data = Vec::new();
+        let mut map = self.map_mutex.lock().unwrap();
+        map.insert(path.clone(), Arc::new(Mutex::new(data)));
+        map.get_mut(&path)
+            .map(|d| InMemoryFile::new(d.clone()))
+            .ok_or("Unreachable".into())
+    }
+
+    fn open_storage(&'a self, path: String) -> Result<InMemoryFile, String> {
+        let mut map = self.map_mutex.lock().unwrap();
+        map.get_mut(&path)
+            .map(|d| InMemoryFile::new(d.clone()))
+            .ok_or("Unreachable".into())
+    }
+}
+
+pub struct InMemoryFile {
+    file_mutex: Arc<Mutex<Vec<u8>>>,
+    position: u64,
+}
+
+impl InMemoryFile {
+    fn new(data: Arc<Mutex<Vec<u8>>>) -> InMemoryFile {
+        InMemoryFile {
+            file_mutex: data,
+            position: 0,
+        }
+    }
+}
+
+impl FileLen for InMemoryFile {
+    fn len(&self) -> Result<u64, String> {
+        let file = self.file_mutex.lock().unwrap();
+        Ok(file.len() as u64)
+    }
+}
+
+impl Read for InMemoryFile {
+    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        let file = self.file_mutex.lock().unwrap();
+        let len = cmp::min(buf.len(), file.len() - self.position as usize);
+        buf[..len].copy_from_slice(&file[self.position as usize..self.position as usize + len]);
+        self.position += len as u64;
+        Ok(len)
+    }
+}
+
+impl Write for InMemoryFile {
+    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
+        let mut file = self.file_mutex.lock().unwrap();
+        file.extend_from_slice(buf);
```

Comment:
It should probably overwrite bytes instead, but yeah, that is an issue. Simple solution would be panic/error on writing when file cursor isn't at the end of the file since we don't actually use seeking at the moment, but we may want to be able to do that later down the line for more sophisticated file-updating algorithms.

Reply:
If seek is not currently used then deleting it, or making it panic if the function is required for some interface/similar, is likely the easiest choice here